### ISSUE TITLE
False positive E0308 when iterating and matching enum items from an impl Into

### ISSUE THREAD

**Issue by Enet4** [rust-lang/rust-analyzer#16657]
The following code compiles without issues in the latest stable Rust, but rust-analyzer makes some incorrect type inference assumptions which lead to error E0308.

```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
struct Tag(u32);

#[allow(dead_code)]
enum SelectorStep {
    Bar(Tag),
    Baz { tag: Tag, item: u32 },
}

struct Selector {
    steps: Vec<SelectorStep>,
}

impl Selector {
    fn iter(&self) -> impl Iterator<Item = &SelectorStep> {
        self.steps.iter()
    }
}

#[allow(dead_code)]
fn process_steps(
    selector: impl Into<Selector>,
) {
    let selector = selector.into();
    let mut map = std::collections::HashMap::new();

    for (i, step) in selector.iter().enumerate() {
        match step {
            SelectorStep::Bar(tag) => {
                map.insert(*tag, i);
            },
            SelectorStep::Baz { tag, .. } => {
                let x = map.get(tag);
                println!("Baz {x:?}");
            }
        }
    }
}

fn main() {}
```

```none
expected &{unknown}, found Tag     rust-analyzer(E0308) [Ln 33, Col 33]
```

```json
[{
	"resource": ".../use-clippy-wat/src/main.rs",
	"owner": "rustc",
	"code": {
		"value": "E0308",
		"target": {
			"$mid": 1,
			"external": "https://doc.rust-lang.org/stable/error_codes/E0308.html",
			"path": "/stable/error_codes/E0308.html",
			"scheme": "https",
			"authority": "doc.rust-lang.org"
		}
	},
	"severity": 8,
	"message": "expected &{unknown}, found Tag",
	"source": "rust-analyzer",
	"startLineNumber": 33,
	"startColumn": 33,
	"endLineNumber": 33,
	"endColumn": 36
}]
```

Following the suggestion to add `&` before `tag` in `let x = map.get(tag);` will trigger the `needless_borrow` Clippy warning.

**rust-analyzer version**: both 0.4.1857-standalone and 0.3.1850-standalone reproduce this

**rustc version**: rustc 1.76.0 (07dca489a 2024-02-04)

**relevant settings**: :shrug:

**Comment by flodiebold** [comment]
Duplicate of #5514
